### PR TITLE
LPC55xxx flash driver

### PR DESCRIPTION
**Pull request by agansari**
Adds flash support to lpc55xxx SoC's.

Tested with samples/drivers/flash_shell

### CODE REVIEW

File: drivers/flash/flash_handlers.c [hunk 1]
```diff
@@ -65,7 +65,7 @@ static inline int z_vrfy_flash_get_page_info_by_idx(struct device *dev,
 }
 #include <syscalls/flash_get_page_info_by_idx_mrsh.c>
 
-static inline size_t z_vrfy_flash_get_page_count(struct device *dev);
+static inline size_t z_vrfy_flash_get_page_count(struct device *dev)
```
Comment (galak) [inline drivers/flash/flash_handlers.c]: Can you pull this out into its own PR, seems like a bug fix that would be nice to get into 2.1 if possible.

Comment (agansari) [reply drivers/flash/flash_handlers.c]: @galak moved to #21064

**Comment by d3zd3z** [comment]
"partiotions" mispelled in commit text of "dts: lpc55s69 all flash regions".

File: drivers/flash/soc_flash_mcux.c [hunk 2]
```diff
@@ -89,14 +110,116 @@ static int flash_mcux_write(struct device *dev, off_t offset,
...
+	rc = FLASH_Program(&priv->config, (u32_t)addr, (u8_t *)data, len);
 	irq_unlock(key);
 
 	k_sem_give(&priv->write_lock);
 
 	return (rc == kStatus_Success) ? 0 : -EINVAL;
 }
 
+#ifdef CONFIG_SW_MANAGED_FLASH_WRITE
+static int flash_mcux_write_unsafe(struct device *dev, off_t offset,
```
Comment (d3zd3z) [inline drivers/flash/soc_flash_mcux.c]: When would someone ever want to define this? If power is lost, it can result in drastic corruption of flash. I would argue that the code using the flash driver always needs to know about the write alignment, and this code should just not be present.

File: drivers/flash/Kconfig.mcux [hunk 3]
```diff
@@ -11,3 +11,16 @@ config SOC_FLASH_MCUX
 	  the duration of the flash erase/write operations. This will
 	  have an impact on the overall system performance - whether
 	  this is acceptable or not will depend on the use case.
+
+if SOC_FLASH_MCUX
+
+config SW_MANAGED_FLASH_WRITE
+	bool "MCUX flash software managed write"
+	default y
```
Comment (d3zd3z) [inline drivers/flash/Kconfig.mcux]: I'm very unwilling to allow this code in at all. It certainly should not be a "default y". It really shouldn't be present at all.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
The write unsafe code _must_ be `default n`, at best. I would argue this code shouldn't even be present, but I'd be willing to concede a little and have the code present if someone can present a convincing argument that it is useful.

File: drivers/flash/Kconfig.mcux (hunk 3 above)
Comment (agansari) [reply drivers/flash/Kconfig.mcux]: @d3zd3z i accidentally force pushed this to 'y' with some generated file, will set back to 'n'

File: drivers/flash/soc_flash_mcux.c (hunk 2 above)
Comment (agansari) [reply drivers/flash/soc_flash_mcux.c]: @d3zd3z the point of this code is to have a simple and independent code from file systems that can be used to test flash writing.
_E.g. samples/drivers/flash_shell_
```
uart:~$ flash page_read 180 20
ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff
ff ff ff ff 
uart:~$ 
uart:~$ 
uart:~$ 
uart:~$ flash page_write 180 10 1 2 3
Reading back written bytes:
01 02 03 
uart:~$ 
uart:~$ 
uart:~$ 
uart:~$ flash page_read 180 20
ff ff ff ff | ff ff ff ff
ff ff 01 02 | 03 ff ff ff
ff ff ff ff 
uart:~$ 
```

Without this code writing 512 bytes would not be possible in console, hacking code would be needed to run this sample.
Any flash programming without provisioning can lead to data corruption in case of power loss.
This function can be used as a generic low level flash write for custom sizes, I would move it to _flash.h_ .
I would just the following warnings:
 - use at your own risk
 - should not be used often, may lead to data corruption
 - when using ensure you have a stable power source

File: drivers/flash/Kconfig.mcux [hunk 4]
```diff
@@ -11,3 +11,16 @@ config SOC_FLASH_MCUX
 	  the duration of the flash erase/write operations. This will
 	  have an impact on the overall system performance - whether
 	  this is acceptable or not will depend on the use case.
+
+if SOC_FLASH_MCUX
+
+config SW_MANAGED_FLASH_WRITE
```
Comment (d3zd3z) [inline drivers/flash/Kconfig.mcux]: It is still unclear to me what the use case is here? This should never be defined.

File: drivers/flash/soc_flash_mcux.c (hunk 2 above)
Comment (d3zd3z) [reply drivers/flash/soc_flash_mcux.c]: But what real purpose does it have? I would argue that the flash shell should be updated to understand how flash devices work. We shouldn't be hiding the behavior of the underlying flash device, especially in a manner that results in corruption.

It _definitely_ should not be a general feature of the flash API.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
I'm still going to strongly resist having this unaligned write feature in this change. I've already seen this show up in another context, and someone just enabled it instead of fixing the underlying problem with the code using the flash driver. This shouldn't be easy to enable. Corruption is subtle, and not usually seen during testing, and we shouldn't have features that make it automatically happen.

So, please just remove this. Fix the shell if it has problems with larger write sizes.

**Comment by d3zd3z** [comment]
Some thoughts on what I might consider an acceptable way of doing this misaligned write. It should _never_ happen as a result of calling the normal `flash_write()`. Ever. If we really want this, it should be a separate API. Code calling it needs to know that what it is doing is unsafe, and other code should never get this behavior just because of a configuration change.

Perhaps call the API `flash_unsafe_misaligned_write_you_should_never_ever_call_this()`?

This implementation wouldn't need to be specific to a driver.

But, at this point, why not just put this functionality into the flash shell. But, perhaps the flash shell should even print a warning on misaligned writes.

What about handling rewrites of data? Should the shell handle that as well? That will generate an error on many devices these days.

**Comment by d3zd3z** [comment]
Realistically, the API could be called something like `flash_unsafe_write_unaligned()`. It doesn't really need to be that short of a name, because its use should be uncommon.

**Comment by agansari** [comment]
Removed unaligned sw managed flash write commit in order to get basic driver merged.
Commit will be moved to a separate pull request.

**Review by d3zd3z — APPROVED** [review]
LGTM

**Review by microbuilder — APPROVED** [review]

**Comment by d3zd3z** [comment]
BTW, as far as the flash misalignment changes go, please put those into the flash shell, not as another patch that tries to add them to either this flash driver, or the flash subsystem itself.

The flash drivers and subsystem are working correctly, and should not contain workarounds that reduce reliability for every client, just because one client isn't implemented correctly.

**Review by nvlsianpu — APPROVED** [review]

File: modules/Kconfig.mcux [hunk 5]
```diff
@@ -63,6 +63,12 @@ config HAS_MCUX_FTFX
 	  Set if the flash memory (FTFA, FTFE, or FTFL) module is present in
 	  the SoC.
 
+config HAS_MCUX_IAP
```
Comment (MaureenHelm) [inline modules/Kconfig.mcux]: Please move this down to keep alpha order

File: dts/bindings/flash_controller/nxp,lpc-iap.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+title: NXP LPC In Application Programming Flash Memory (IAP)
```
Comment (MaureenHelm) [inline dts/bindings/flash_controller/nxp,lpc-iap.yaml]: Add license/copyright header

File: drivers/flash/soc_flash_mcux.c [hunk 7]
```diff
@@ -14,7 +14,11 @@
 #include "flash_priv.h"
 
 #include "fsl_common.h"
+#ifdef CONFIG_SOC_FAMILY_LPC
```
Comment (MaureenHelm) [inline drivers/flash/soc_flash_mcux.c]: It would be more consistent with the Kconfig dependency to use `CONFIG_HAS_MCUX_IAP` here

File: drivers/flash/soc_flash_mcux.c [hunk 8]
```diff
@@ -149,8 +157,13 @@ static int flash_mcux_init(struct device *dev)
 
 	rc = FLASH_Init(&priv->config);
 
+#ifdef CONFIG_SOC_FAMILY_LPC
```
Comment (MaureenHelm) [inline drivers/flash/soc_flash_mcux.c]: Same comment as above regarding `CONFIG_HAS_MCUX_IAP`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by agansari** [comment]
Thanks @MaureenHelm I've updated and made another pull.

@d3zd3z the flash_shell was just one example where I needed unaligned writes; I can think of other examples such as logging applications that work with small data.
I find there is more benefit than risk, in having platform agnostic, alignment independent and easy to use write functions.

**Comment by d3zd3z** [comment]
> I find there is more benefit than risk, in having platform agnostic, alignment independent and easy to use write functions.

There is just no way to do this without making whatever system is doing it fail hopelessly when power is lost. Code that uses flash *must* know the characteristics of that flash device. Things like logging should be using a filesystem on top of a flash device, not the flash device itself.

I guess I don't really have a problem with a library like this (I don't think it should have any clients), but it is very important to me that the flash drivers themselves not be broken, and only support operations that can always be done with known characteristics.

Flash devices are hard to use. That's where there are filesystems that work on top of them. It doesn't help anyone to have things that might look like they make it easy to use, but only result in corruption and loss of data.

**Review by MaureenHelm — APPROVED** [review]
